src/ir/bits.h

```diff
@@ -128,35 +128,111 @@ struct DummyLocalInfoProvider {
 template<typename LocalInfoProvider = DummyLocalInfoProvider>
 Index getMaxBits(Expression* curr,
                  LocalInfoProvider* localInfoProvider = nullptr) {
-  if (auto* const_ = curr->dynCast<Const>()) {
+  if (auto* c = curr->dynCast<Const>()) {
     switch (curr->type.getSingle()) {
       case Type::i32:
-        return 32 - const_->value.countLeadingZeroes().geti32();
+        return 32 - c->value.countLeadingZeroes().geti32();
       case Type::i64:
-        return 64 - const_->value.countLeadingZeroes().geti64();
+        return 64 - c->value.countLeadingZeroes().geti64();
       default:
         WASM_UNREACHABLE("invalid type");
     }
   } else if (auto* binary = curr->dynCast<Binary>()) {
     switch (binary->op) {
       // 32-bit
-      case AddInt32:
-      case SubInt32:
-      case MulInt32:
-      case DivSInt32:
-      case DivUInt32:
-      case RemSInt32:
-      case RemUInt32:
       case RotLInt32:
       case RotRInt32:
         return 32;
-      case AndInt32:
-        return std::min(getMaxBits(binary->left, localInfoProvider),
-                        getMaxBits(binary->right, localInfoProvider));
+      case AddInt32: {
+        auto maxBitsLeft = getMaxBits(binary->left, localInfoProvider);
+        auto maxBitsRight = getMaxBits(binary->right, localInfoProvider);
+        return std::min(Index(32), std::max(maxBitsLeft, maxBitsRight) + 1);
+      }
+      case SubInt32: {
+        auto maxBitsLeft = getMaxBits(binary->left, localInfoProvider);
+        if (maxBitsLeft == 32 || maxBitsLeft == 0) {
+          return 32;
+        }
+        auto maxBitsRight = getMaxBits(binary->right, localInfoProvider);
+        if (maxBitsLeft <= maxBitsRight) {
+          return 32;
+        }
+        return std::max(maxBitsLeft, maxBitsRight);
```

Comment:
I don't think this is right. `getMaxBits` returns an **upper bound** on the bits, but not the exact amount.

For example, we may know that maxBitsLeft is `5` and maxBitsRight is `10`. With those values, we would reach this line with the return, and return 10. But imagine that the actual values at runtime are `0` on the left and `1` on the right. Those values are consistent with the `5`, `10` from earlier, we don't have more than those amounts of bits, but the result is `0 - 1 = -1` which has 32 bits set.

For subtraction,  maybe the simplest way to think about it is with `x - y == x + (-y)`. That is, negate the right hand, and then it's a simple add. The problem is that negation doesn't allow us to do an easy computation of getMaxBits, I don't think.

Reply:
> For example, we may know that maxBitsLeft is 5 and maxBitsRight is 10. With those values, we would reach this line with the return, and return 10.
```cpp
if (maxBitsLeft <= maxBitsRight) {
  return 32
}
```
will return in this case `32`. Also it return 32 even when `maxBitsLeft == maxBitsRight`. So only when `maxBitsRight < maxBitsLeft` we calculate max(maxBitsRight, maxBitsLeft). Btw in this case we could just return `maxBitsLeft` instead get max of lhs and rhs.

[thare related tests](https://github.com/WebAssembly/binaryen/pull/2879/files#diff-74356d0dc45bf43c3e3b1e6bc15be202R84)

Reply:
Added [`0 - 1` test](https://github.com/WebAssembly/binaryen/pull/2879/commits/df40a08c1b160c3b1b97e3b4e4e12bcdaab613df)

Reply:
The tricky thing here is the difference between the runtime values and the values estimated on the wasm. Here is a more complete example:
```wat
(func $sub_with_masks (param $x i32) (param $y i32) (result i32)
 (i32.sub
  (i32.and (local.get $x) (i32.const 255))
  (i32.and (local.get $y) (i32.const 127))
 )
)
(func $test (result i32)
  (call $sub_with_masks (i32.const 0) (i32.const 1))
)
```
`getMaxBits` returns `8` on the left and `7` on the right. But the **runtime** values will be `0` and `1`, so we will be computing `0 - 1` and returning `-1` which has `32` bits used. (For the bug to be noticed, the first function would need to be inlinined into a place where `getMaxBits` actually affects optimization, etc., so the fuzzer can't quickly find counterexamples, I am guessing.)

`getMaxBits` just says that the value has less than that amount of bits, or less. Since it can have less, if X has more maximum bits than Y, it can still have less bits in practice at runtime.

Reply:
Hmm, it seems this example also not affected. I added example in [recent commit](https://github.com/WebAssembly/binaryen/pull/2879/commits/da177ae35deb89755d347e84f068afd002f5b103). So if `getMaxBits` for `$sub_with_masks` calculate as `8` bits it should elleminate `(i32.and (call $sub_with_masks 0 1) 1023)` due to `10 > 8` and masking by 1023 unnecessary. But it's not happening, so I guess it already handled by LocalInfo provider which know much more precisely. But if this local comes from import it could be a problem I guess

Reply:
The example in that commit isn't affected because we don't propagate this information between functions. (As I said earlier, for the bug to hurt us there would need to be inlining, or more code around it, which is why the fuzzer is unlikely to hit it.)

But here's a full example, that doesn't need further inlining:
```wat
(func $stuff_with_masks (param $x i32) (param $y i32) (result i32)
 (i32.and
  (i32.const 0xff)
  (i32.sub
   (i32.and (local.get $x) (i32.const 255))
   (i32.and (local.get $y) (i32.const 127))
  )
 )
)
(func $test (result i32)
  (call $sub_with_masks (i32.const 0) (i32.const 1))
)
```
Your PR calculates `getMaxBits` of the `sub` as 8, because left has at most 8, right has at most 7, and therefore `maxBitsLeft <= maxBitsRight` is false. We then apply an `and` mask of `0xff`. But as the max bits is 8, masking to leave just the lower 8 bits is redundant, and the pass will remove the `and` (that's existing code in the pass from before this PR, see the comment `a mask of lower bits is not needed if we are already smaller`). But that leads to incorrect results, as `$test` shows - it will return `0xffffffff` (since the mask hasn't been done) instead of `0xff`.

Reply:
Now I see. Thanks for explanation! I'll revert this back to previous form which just return max possible bits for current type.

Reply:
Done!